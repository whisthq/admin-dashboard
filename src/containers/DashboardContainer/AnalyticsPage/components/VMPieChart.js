import React, { Component } from 'react'
import { connect } from 'react-redux'
import { PieChart, Pie, ResponsiveContainer, Cell } from 'recharts'

import '../../../../static/App.css'

class VMPieChart extends Component {
    render() {
        const data = [
            { name: 'Deallocated', value: this.props.deallocated },
            { name: 'Running Available', value: this.props.available },
            { name: 'Running Unavailable', value: this.props.unavailable },
        ]
        const colors = ['#bf3762', '#15d157', '#e2e620']
        return (
            <div>
                <ResponsiveContainer width={200} height={200}>
                    <PieChart>
                        <text
                            x={100}
                            y={85}
                            textAnchor="middle"
                            dominantBaseline="middle"
                            style={{
                                fontSize: 40,
                                fontWeight: 'bold',
                            }}
                        >
                            {this.props.deallocated +
                                this.props.available +
                                this.props.unavailable}
                        </text>
                        <text
                            x={100}
                            y={125}
                            textAnchor="middle"
                            dominantBaseline="middle"
                            style={{
                                fontSize: 14,
                                fontWeight: 'normal',
                                color: '#777777',
                            }}
                        >
                            {this.props.location}
                        </text>
                        <Pie data={data} dataKey="value" innerRadius={70}>
                            {data.map((entry, index) => (
                                <Cell fill={colors[index % colors.length]} />
                            ))}
                        </Pie>
                    </PieChart>
                </ResponsiveContainer>
            </div>
        )
    }
}

export default connect()(VMPieChart)
