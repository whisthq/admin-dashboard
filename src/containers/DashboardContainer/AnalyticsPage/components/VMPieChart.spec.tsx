import React from 'react'
import Enzyme, { shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { PieChart, Pie, Cell } from 'recharts'

import { VMPieChart } from './VMPieChart'

// necessary for enzyme to work
Enzyme.configure({ adapter: new Adapter() })
/*
The VMPieChart uses ReCharts do siplay pretty pie chart circle edges with colors failure or success.

It should have a cell per color in a pie in a pie chart. It should take in props for deallocated, available,
unavailable, and location and then output their sum as the text above the location as well as a cell w/ color
per each of these props.
*/
describe('<VMPieChart />', () => {
    it('allows us to set props and will display correctly, displays both name and sum', () => {

        const wrapper = mount(<VMPieChart deallocated={3} available={2} unavailable={1} location="foo"/>)
        expect(wrapper.props().deallocated).toEqual(3)
        expect(wrapper.props().available).toEqual(2)
        expect(wrapper.props().unavailable).toEqual(1)
        expect(wrapper.props().location).toEqual("foo")

        expect(
            wrapper.containsAllMatchingElements([
                <text>6</text>, //sum of deallocated, available, and unavailable
                <text>foo</text>,
            ])
        ).toEqual(true)
    })

    it('has a pie in a pie chart with a cells per color', () => {
        const numColors = 3

        const wrapper = shallow(<VMPieChart />)
        expect(wrapper.find(PieChart)).toHaveLength(1)
        expect(wrapper.find(Pie)).toHaveLength(1)
        expect(wrapper.find(Cell)).toHaveLength(numColors)
    })

    it('maps each entry in data to a corresponding cell in a pie', () => {
        const colors = ['#bf3762', '#15d157', '#e2e620']
        const colorsRemaining = new Set(colors)

        const wrapper = mount(
            <VMPieChart
                deallocated={0}
                available={0}
                unavailable={0}
                location=""
             />
        )
        const foundCells = wrapper.find(Cell) // we checked in a previous test for the length
        
        foundCells.forEach((cell) => {
            let color = cell.prop("fill")
            color = color ? color : "" // undefined -> ""

            expect(colorsRemaining.has(color)).toEqual(true)
            colorsRemaining.delete(color)
        })

    })
})