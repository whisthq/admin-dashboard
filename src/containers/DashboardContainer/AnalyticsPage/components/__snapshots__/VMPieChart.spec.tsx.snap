// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VMPieChart /> renders correctly 1`] = `
<div>
  <ResponsiveContainer
    debounce={0}
    height={200}
    width={200}
  >
    <PieChart
      barCategoryGap="10%"
      barGap={4}
      cx="50%"
      cy="50%"
      endAngle={360}
      innerRadius={0}
      layout="centric"
      margin={
        Object {
          "bottom": 5,
          "left": 5,
          "right": 5,
          "top": 5,
        }
      }
      outerRadius="80%"
      reverseStackOrder={false}
      stackOffset="none"
      startAngle={0}
    >
      <text
        dominantBaseline="middle"
        style={
          Object {
            "fontSize": 40,
            "fontWeight": "bold",
          }
        }
        textAnchor="middle"
        x={100}
        y={85}
      >
        6
      </text>
      <text
        dominantBaseline="middle"
        style={
          Object {
            "color": "#777777",
            "fontSize": 14,
            "fontWeight": "normal",
          }
        }
        textAnchor="middle"
        x={100}
        y={125}
      >
        foo
      </text>
      <Pie
        animationBegin={400}
        animationDuration={1500}
        animationEasing="ease"
        blendStroke={false}
        cx="50%"
        cy="50%"
        data={
          Array [
            Object {
              "name": "Deallocated",
              "value": 3,
            },
            Object {
              "name": "Running Available",
              "value": 2,
            },
            Object {
              "name": "Running Unavailable",
              "value": 1,
            },
          ]
        }
        dataKey="value"
        endAngle={360}
        fill="#808080"
        hide={false}
        innerRadius={70}
        isAnimationActive={true}
        labelLine={true}
        legendType="rect"
        minAngle={0}
        nameKey="name"
        outerRadius="80%"
        paddingAngle={0}
        startAngle={0}
        stroke="#fff"
      >
        <Cell
          fill="#bf3762"
          key="0"
        />
        <Cell
          fill="#15d157"
          key="1"
        />
        <Cell
          fill="#e2e620"
          key="2"
        />
      </Pie>
    </PieChart>
  </ResponsiveContainer>
</div>
`;
